xdsl/irdl/declarative_assembly_format_parser.py

```diff
@@ -487,12 +487,21 @@ def parse_optional_group(self) -> FormatDirective:
                 anchor = then_elements[-1]
         self.parse_punctuation("?")
 
-        if not then_elements:
+        # Pull whitespace element of front, as they are not parsed
+        whitespace_counter = 0
+        while whitespace_counter < len(then_elements) and isinstance(
+            then_elements[whitespace_counter], WhitespaceDirective
+        ):
+            whitespace_counter += 1
```

Comment:
Here's a maybe Pythonic way?
```suggestion
        whitespace_counter = next(
            (
                i
                for i, e in enumerate(then_elements)
                if not isinstance(e, WhitespaceDirective)
            ),
            len(then_elements),
        )
```
I don't find it necessarily clearer myself, but answering as requested!

Reply:
I think I personally find it less clear what this does, but happy to go either way

Reply:
> I think I personally find it less clear what this does, but happy to go either way

Yeah I don't mind myself; as long as it works and is understandable - I'd say both styles are both for the everyday xDSL dev!

Reply:
if all the elements are whitespace then it should throw an error (as the next part of the code does)

Reply:
Yep this one kept it, you get the length of the list out so you would trigger the exception down

Otherwise, just remove the len(then_elements) from there and next() itself would throw an error, but not a nice hand-written one..

You could replace len(then_elements) in there by None and just
```python
if whitespace_counter is None:
    # fancy_error
```

Reply:
I would recommend something like this:
```suggestion
        first_non_whitespace_index = 0
        for first_non_whitespace_index, el in enumerate(then_elements):
            if isinstance(el, WhitespaceDirective):
                break
```